### DISCUSSION TITLE
Set value for multiselect from state (material-ui select)

### DISCUSSION THREAD

**Discussion by ngurahyudi** [react-hook-form/react-hook-form#5405] [Q&A] [answered]
Hello, I have a custom component (select with dynamic checkboxes include select all checkbox), I want to set the value based on state because when user check the select all option i can programmatically set the state (create string array based on checkbox values but select all checkbox value). Here is my code snapshot. 

![code jpg](https://user-images.githubusercontent.com/16011522/120065400-b099be00-c0a3-11eb-87bb-9a51b402ebb2.png)

The problem is when i check the select all checkbox and console log the data, it shows ['all'] which is the value of the checkbox
its seems the select value not returning my selected state.

Any advice ?
Thank you in advance

**Comment by thanh-nguyen-95** [comment, marked as answer]
You have to call both `onChange` (to update `react-hook-form`'s values) and `setState` (to update your own state) in all cases

For example:

```jsx
<Select
  onChange={(event: ChangeEvent<HTMLSelectElement>) => {
    const { target: { value } } = event

    if (isAllOptionsSelected) {
      const data= /* handle data transform here */
      onChange(data)
      setState(data)
    } else {
      onChange(value)
      setState(value)
    }
  }}
/>
```

**Side notes:** Please provide either a CodeSandbox (recommended) or a code snippet (like my example) for better readability and easier support from others.

Cheers!

**Reply by ngurahyudi** [reply]
Hi nguyen, thanks for your answer, but I already do that. Please see at my code at Select `onChange` event. I'll try to provide the CodeSandbox

**Reply by thanh-nguyen-95** [reply]
What you did was calling `onChange` on `event`, which will only get the `['all']` when you select 'all'. My code was to transform the data first, then call `onChange` with the transformed data.

I don't think it's the same. Kindly check the solution again.

**Reply by ngurahyudi** [reply]
OMG It's works! I've been struggling on this issue for hours!
Thank You, Nguyen you're a life saver! :)

**Reply by ngurahyudi** [reply]
Hi Nguyen, sorry for open this topic again, now I also want to reset the selected state when rhf reset is called, how can i do that?
this is my CSB link :
https://codesandbox.io/s/upbeat-river-xxpy8

**Reply by thanh-nguyen-95** [reply]
> 
> 
> Hi Nguyen, sorry for open this topic again, now I also want to reset the selected state when rhf reset is called, how can i do that?
> this is my CSB link :
> https://codesandbox.io/s/upbeat-river-xxpy8

I apologize for the late response. I've been very busy lately.

To reset the value of a component, it should be a controlled component. Your select is controlled (since it has both `value` and `onChange` props) but not controlled correctly.

The value from the select is taken from `useState`, not `react-hook-form` itself, which make `reset` not working as intended.

The solution is to change the select component as below:

```jsx
<Select
  ref={ref} // as 'field.ref'
  name={name} // as 'field.name'
  value={value} // as 'field.value' previously 'value={selected}'
  onChange={(e) => {
    const value = e.target.value;
    if (value[value.length - 1] === "all") {
      const sel =
        selected.length === props.data.length
          ? []
          : props.data.map((e) => e.value);
      onChange(sel);
      setSelected(sel);
      return;
    }
    onChange(value);
    setSelected(value);
  }}
  /* other props here */
>
  /* render options here */
</Select>
```

It's important to pass `ref`, `name`, `value`, `onChange` to your component when using `react-hook-form`. You should also consider removing `selected` state inside the component and using `props.value` as values instead.

**Reply by bluebill1049** [reply]
thanks @thanh-nguyen-95

**Reply by ngurahyudi** [reply]
@thanh-nguyen-95 thank you for your clear explanation. I'm very new to this library and it's help me a lot @bluebill1049 thank you. I'll come soon when I get a new question and I hope you don't mind.
